### PR TITLE
add bls crypto functions to transaction and script env

### PR DESCRIPTION
**Pull request by dsainati1**
In order to address https://github.com/onflow/cadence/issues/1127, https://github.com/onflow/cadence/issues/1128 and https://github.com/onflow/cadence/issues/1129, we need to add the relevant functions to `scriptEnv.go` and `transactionEnv.go` 

The Cadence portion of this work is in https://github.com/onflow/cadence/pull/1236

This adds functions to these environments that delegate to the proper crypto library.

### CODE REVIEW

File: crypto/sign_norelic.go
Comment (janezpodhostnik) [inline crypto/sign_norelic.go:2]: Not sure why both are needed. But my IDE(goland) has a problem with both while just `// +build !relic` works as expected.

**Review by janezpodhostnik — COMMENTED** [review]
Will the cadence version update where the `runtime.Interface` is updated to include this methods be included at a later date?

Usually I included the cadence version update in a PR like this. This allowed me to write tests like this one: https://github.com/onflow/flow-go/blob/192cd839169d49207b4e2681b7d96cfc86f84f90/fvm/fvm_test.go#L1380.

**Comment by dsainati1** [comment]
> Will the cadence version update where the runtime.Interface is updated to include this methods be included at a later date?

I have local changes to the interface on the Cadence side, but was not aware that it was possible to include changes to both repositories in a single PR. I will look into getting those changes into this PR.

File: crypto/sign_norelic.go
Comment (dsainati1) [reply crypto/sign_norelic.go:2]: Hmm it seems like this was automatically added by VSCode.

File: crypto/sign_norelic.go
Comment (tarakby) [inline crypto/sign_norelic.go:28]: In the crypto package, these function should not be exported when Relic tag is not enabled.
I understand the need in FVM is different (these functions should always be defined). Since it is an FMV issue, I suggest we define these functions (relic and non-relic) in `fvm/crypto`. This is already done for BLS KMAC ([relic](https://github.com/onflow/flow-go/blob/cadence-issue-1127/fvm/crypto/blskmac_relic.go) and [non-relic](https://github.com/onflow/flow-go/blob/cadence-issue-1127/fvm/crypto/blskmac.go)).
The files can be renamed to reflect that they now contain multiple crypto tools and not just BLS_KMAC.

File: crypto/sign_relic.go
Comment (tarakby) [inline crypto/sign_relic.go:61]: As mentioned above, these functions could also be moved under `fvm/crypto`.

File: fvm/scriptEnv.go
Comment (tarakby) [inline fvm/scriptEnv.go:691]: Not sure how the upper layers of Cadence would handle returned errors. But I know we would like to differentiate 2 kind of errors:
 - recoverable errors: the crypto package returns a sentinel error if any input is invalid (caller mistake) and the node shouldn't crash.
 - non recoverable errors: an unexpected exception during execution, that can cause the node to crash.
 
 This is done here for instance https://github.com/onflow/flow-go/blob/cadence-issue-1127/fvm/crypto/crypto.go#L104-L109.
 
 cc @janezpodhostnik

File: fvm/scriptEnv.go
Comment (tarakby) [inline fvm/scriptEnv.go:695]: Same here, we probably should check for the sentinel error (we can also make the check under fvm/crypto)

File: fvm/scriptEnv.go
Comment (tarakby) [inline fvm/scriptEnv.go:699]: Same here, we probably should check for the sentinel error (we can make the check under fvm/crypto)

File: fvm/scriptEnv.go
Comment (tarakby) [inline fvm/scriptEnv.go:705]: Same here, we probably should check for the sentinel error

File: fvm/scriptEnv.go
Comment (tarakby) [inline fvm/scriptEnv.go:711]: Same here, we probably should check for the sentinel error (we can make the check under fvm/crypto)

File: fvm/transactionEnv.go
Comment (tarakby) [inline fvm/transactionEnv.go:955]: These methods seem to be redundant with `ScriptEnv` methods, it would be good to factorize this code to make it use the same code (better maintainability).

**Review by tarakby — COMMENTED** [review]
Thanks for the updates, it's nice to see these features coming soon to Cadence 🏄🏼 

I have left a few comments below.

File: fvm/scriptEnv.go
Comment (dsainati1) [reply fvm/scriptEnv.go:691]: > recoverable errors: the crypto package returns a sentinel error if any input is invalid (caller mistake) and the node shouldn't crash.

Would you want these functions to return `nil` on an invalid input?

**Comment by dsainati1** [comment]
I have put up a PR for the Cadence side of this work in https://github.com/onflow/cadence/pull/1236

In order to test these changes, I needed to update to use my Cadence changes, which meant I needed to rebase this PR onto the `v0.23-testnet` branch. I rebased over https://github.com/onflow/flow-go/pull/1613 to fix some build errors.

File: fvm/transactionEnv.go
Comment (ramtinms) [inline fvm/transactionEnv.go:1011]: I wonder if we can prefix all of these methods with BLS and then add the method names (e.g. `BLSAggregateSignatures`) so we naturally have some grouping of these methods,

**Review by ramtinms — APPROVED** [review]
Looks good to me, just make sure we port this over to the master as well.

File: fvm/fvm_test.go
Comment (tarakby) [inline fvm/fvm_test.go:2053]: Nice solution! 
Another (faster) way would be to aggregate the signatures straight away.

```suggestion
							expectedSig, err := crypto.AggregateBLSSignatures(sigs)
							require.NoError(t, err)
```

Comment (tarakby) [reply fvm/fvm_test.go:2053]: Well, I figured we will need to construct `sigs` too.

File: fvm/fvm_test.go
Comment (tarakby) [inline fvm/fvm_test.go:2058]: Here too we should be testing `script.Value` against Cadence default value (`nil` maybe?)

File: fvm/fvm_test.go
Comment (tarakby) [inline fvm/fvm_test.go:2136]: What is implemented is correct, though we can make a shortcut. 
The tests in the crypto library are made differently because they are testing the arithmetic relationships, while here we're just testing the FVM functions are wired correctly.
```suggestion
							expectedPk, err := crypto.AggregateBLSPublicKeys(sks)
							require.NoError(t, err)
```

File: fvm/fvm_test.go
Comment (tarakby) [inline fvm/fvm_test.go:2140]: Same comment about testing the `script.value` against Cadence default vaule.

**Review by tarakby — COMMENTED** [review]
Thanks for adding the tests and making the changes!
I left a few comments about the tests, most of them are minor.  

- I would like to add a few tests (making a PR against yours), though I am having issues importing Cadence package to test locally:
```
go: github.com/onflow/cadence@v0.20.0-beta10: reading github.com/dapperlabs/cadence-internal/go.mod at revision v0.20.0-beta10-patch.1: unknown revision v0.20.0-beta10-patch.1
```

- Regarding testing against the crypto sentinel error for invalid inputs, I answer your question here:
> Would you want these functions to return nil on an invalid input?

Yes, it should be `false` for `verifyPoP` and `nil` for the others.

**Review by janezpodhostnik — COMMENTED** [review]
Besides the comments that Tarak has, the res looks good to me.

I agree that incorrect use input into `BLSVerifyPOP` should not produce an error on cadence side, but should just return false.

**Comment by dsainati1** [comment]
> I would like to add a few tests (making a PR against yours), though I am having issues importing Cadence package to test locally:

I just used the master build of cadence, replacing Cadence in the go.mod file: `go mod edit -replace github.com/onflow/cadence=../cadence`

**Comment by tarakby** [comment]
Here is the PR extending the tests : https://github.com/onflow/flow-go/pull/1639 
Please feel free to review it and we can merge it into your branch.

**Comment by tarakby** [comment]
While we figure out the error return types to Cadence to differentiate user's errors and unexpected execution errors, I would like to add another specification. 
I am not sure if it should be added to the FVM (this PR) or the runtime (https://github.com/onflow/cadence/pull/1236) though:

In the `PublicKey` structure in Cadence, there is a field called `isValid`, that is assigned to true or false when the structure is created (Cadence doc: https://docs.onflow.org/cadence/language/crypto/). You can notice that when the field is set to false, and the function `verify` is called on the public key, the function returns false. Similarly: 
 - `verifyPoP` should return false when the public key field `isValid` is false.
 - `BLSAggregatePublicKeys` should return nil (not sure whether it needs to error too) if at least one input public key has a false `isValid` field.

File: go.mod
Comment (turbolent) [inline go.mod:97]: This change for local development purposes should not be committed

**Comment by dsainati1** [comment]
> While we figure out the error return types to Cadence to differentiate user's errors and unexpected execution errors, I would like to add another specification. I am not sure if it should be added to the FVM (this PR) or the runtime ([onflow/cadence#1236](https://github.com/onflow/cadence/pull/1236)) though:

I have added these changes to https://github.com/onflow/cadence/pull/1250 on the Cadence side. The condition is easy to check just using the `PublicKey` values and abort execution if they are invalid, so no need to call into the FVM.

**Review by janezpodhostnik — APPROVED** [review]
The error handling looks like what we discussed, so LGTM.

**Comment by dsainati1** [comment]
This is now waiting on https://github.com/onflow/cadence/pull/1250 to add the corresponding changes to the Cadence side, which is in turn blocked on https://github.com/onflow/cadence/pull/1273. Once these two PRs are merged and included in the next Cadence release, we should be able to merge this.

**Comment by tarakby** [comment]
This PR is awaiting a PR to update Cadence, being worked on by @janezpodhostnik.

**Comment by dsainati1** [comment]
We are also planning on releasing the BLS changes to Cadence in version 0.23

**Comment by janezpodhostnik** [comment]
> This PR is awaiting a PR to update Cadence, being worked on by @janezpodhostnik.

PR https://github.com/onflow/flow-go/pull/1975 is ready, except the crypto changes. But those are actually here. It might make sense to either base https://github.com/onflow/flow-go/pull/1975 on this one or the other way around. What would you prefer?

**Comment by turbolent** [comment]
Updated this PR to the latest version of Cadence, v0.23.0, which we would like to get into the next spork.

There is one remaining ytes

**Comment by turbolent** [comment]
@tarakby Could you please double check [fe42f9d](https://github.com/onflow/flow-go/pull/1605/commits/fe42f9dc85c8de91579ea2b8d7836a7209c70dc6)?

File: fvm/bootstrap.go
Comment (turbolent) [inline fvm/bootstrap.go:682]: Was this change necessary? It shouldn't be

**Review by turbolent — APPROVED** [review]
Looks good! Just one question regarding the Cadence code-change

File: fvm/bootstrap.go
Comment (janezpodhostnik) [reply fvm/bootstrap.go:682]: Might not be for cadence 0.23. Let me check

Comment (janezpodhostnik) [reply fvm/bootstrap.go]: removed [6d883fd](https://github.com/onflow/flow-go/pull/1605/commits/6d883fd26a95beabcc6c198247ce1e7d1edef6c4)

**Comment by tarakby** [comment]
> @tarakby Could you please double check [fe42f9d](https://github.com/onflow/flow-go/pull/1605/commits/fe42f9dc85c8de91579ea2b8d7836a7209c70dc6)?

Thanks, I'm gonna make a minor change.

File: fvm/crypto/crypto.go
Comment (turbolent) [inline fvm/crypto/crypto.go:106]: Good idea to refactor this also here! 👍

**Review by turbolent — APPROVED** [review]

**Pull request merged**
